src/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/SymbolTable.cs

```diff
@@ -906,40 +823,41 @@ private static List<object> BuildDeclarationChain(Type callingType)
                     Debug.Assert(bAdded);
                 }
             }
+
             callChain.Reverse();
 
             // Now take out the namespaces and add them to the end of the chain.
-
             if (callingType.Namespace != null)
             {
-                string[] namespaces = callingType.Namespace.Split('.');
-                int index = 0;
-                foreach (string s in namespaces)
-                {
-                    callChain.Insert(index++, s);
-                }
+                callChain.InsertRange(0, callingType.Namespace.Split('.'));
             }
             return callChain;
         }
 
-        /////////////////////////////////////////////////////////////////////////////////
+        // We have an aggregate symbol of the correct parent and full name, but it may have the wrong arity, or, due to
+        // dynamic loading or creation, two different types can exist that have the same name.
+
+        // In the static compiler, this would have been an error and name lookup would be ambiguous, but here we never have
+        // to lookup names of types for real (only names of members).
```

Comment:
You are right, in the dynamic binder we never need to lookup a type while having only its name. We always start either from a type or form an instance, which has a type.
The name ambiguity can be trivially resolved.

Reply:
And yet I trusted the comment enough to consider several possible different approaches before realising I could just do it by not trying not to do it! 😄

Reply:
If there was something like a dynamic `new` which creates instances by type name, and dynamic way to introduce types, then this could get more interesting. 
But then name clashes would just be ordinary ambiguities, and there are no such things anyways. :-)